funsor/ops.py

```diff
@@ -15,17 +17,30 @@
 
 
 class Op(namedtuple('Op', ['fn'])):
+    keyed_registry = KeyedRegistry()
+
     def __new__(cls, fn):
         result = super(Op, cls).__new__(cls, fn)
+        # register as default op impl for object args
         functools.update_wrapper(result, fn)
+        args, vargs, kwargs, default = getargspec(fn)
+        assert not vargs
+        assert not kwargs
+        default_signature = (object,) * len(args)
+        cls.keyed_registry.register_impl(result, lambda *args: fn(*args), *default_signature)
```

Comment:
It seems simpler to use a `Dispatcher` object rather than a `KeyedRegistry`, something like
```py
def __new__(cls, fn):
    if not isintance(fn, multipledispatch.Dispatcher):
        old_fn, fn = fn, multipledispatch.Dispatcher(fn.__name__)
        args, vargs, kwargs, default = getargspec(fn)
        default_signature = (object,) * len(args)
        fn.register(*default_signature)(old_fn)
    result = super(Op, cls).__new__(cls, fn)
    functools.update_wrapper(result, fn)
    return result

@property
def register(self):
    return self.fn.register
```

Reply:
Actually, why don't we simply let `Op` inherit from `Dispatcher` and bake your nice `default_signature` logic into `Op.__init__()`? I don't think there was any particular reason to keep `Op` a namedtuple.

Reply:
That really cleaned things up, thanks for the suggestion!